Sample Apps/Storefront/Storefront/ParallaxViewController.swift

```diff
@@ -0,0 +1,153 @@
+    
+    lazy private var proxyView: ProxyView = {
+        let view = ProxyView(frame: self.view.bounds)
+        self.view.addSubview(view)
+        return view
+    }()
+    
+    private var topY: CGFloat {
+        return max(self.insets.top, self.topLayoutGuide.length)
+    }
+    
+    private var midY: CGFloat {
+        return self.topY + self.headerHeight
+    }
+    
+    private var bottomSpace: CGFloat {
+        return self.bottomLayoutGuide.length
+    }
+    
+    // ----------------------------------
+    //  MARK: - View Loading -
+    //
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        
+        self.configureProxyView()
+        self.configureScrollView()
+    }
+    
+    private func configureProxyView() {
+        self.proxyView.headerView = headerView
+        self.proxyView.scrollView = scrollView
+    }
+    
+    private func configureScrollView() {
+        self.scrollView.backgroundColor = .clear
+    }
+    
+    // ----------------------------------
+    //  MARK: - Updates -
+    //
+    func updateParallax() {
+        
+        let headerOffset = self.scrollView.contentOffset.y + self.midY
+        let headerY      = min(self.topY - (headerOffset * self.multiplier), self.topY)
+        let headerHeight = max(self.headerHeight, self.headerHeight - headerOffset)
+        
+        var frame             = self.headerView.frame
+        frame.origin.y        = headerY
+        frame.size.height     = headerHeight
+        self.headerView.frame = frame
+    }
+    
+    // ----------------------------------
+    //  MARK: - Layout Subviews -
+    //
+    override func viewWillLayoutSubviews() {
+        super.viewWillLayoutSubviews()
+        
+        self.layoutProxyView()
+        self.layoutHeaderView()
+        self.layoutScrollView()
+        
+        self.adjustZIndex()
+        self.updateParallax()
+    }
+    
+    private func layoutProxyView() {
+        self.proxyView.frame = self.view.bounds
+    }
+    
+    private func layoutHeaderView() {
+        var frame             = self.view.bounds
+        frame.origin.y        = self.topY
+        frame.size.height     = self.headerHeight
+        self.headerView.frame = frame
+    }
+    
+    private func layoutScrollView() {
+        self.scrollView.frame = self.view.bounds
+        
+        var insets                   = self.scrollView.contentInset
+        insets.top                   = self.midY
+        insets.bottom                = self.bottomSpace
+        self.scrollView.contentInset = insets
+    }
+    
+    private func adjustZIndex() {
+        self.view.insertSubview(self.headerView, at: 0)
+        self.view.insertSubview(self.scrollView, at: 1)
+        self.view.insertSubview(self.proxyView, at: 2)
+    }
+}
+
+// ----------------------------------
+//  MARK: - ProxyView -
+//
+private class ProxyView: UIView {
+    
+    weak var headerView: UIView!
+    weak var scrollView: UIScrollView!
```

Comment:
I assume you mean the `ParallaxViewController` and not `ProxyViewController`.

Why can't it ensure that both are available in `viewDidLoad`? `viewDidLoad` is called after the views have been loaded, either via `loadView` or coming from a nib.

> They can be set at any time and can remain nil prior to being set

If that's the case then they should be weak optionals and not weak implicitly unwrapped because they could be set to being nil at any time.